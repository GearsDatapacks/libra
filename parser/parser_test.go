package parser_test

import (
	"testing"

	"github.com/gearsdatapacks/libra/lexer"
	"github.com/gearsdatapacks/libra/lexer/token"
	"github.com/gearsdatapacks/libra/parser"
	"github.com/gearsdatapacks/libra/parser/ast"
	utils "github.com/gearsdatapacks/libra/test_utils"
)

func getProgram(t *testing.T, input string) *ast.Program {
	l := lexer.New(input, "test.lb")
	tokens := l.Tokenise()

	p := parser.New(tokens, l.Diagnostics)
	program := p.Parse()
	utils.AssertEq(t, len(p.Diagnostics.Diagnostics), 0)

	return program
}

func getExpr[T ast.Expression](t *testing.T, program *ast.Program) T {
	utils.AssertEq(t, len(program.Statements), 1)
	stmt, ok := program.Statements[0].(*ast.ExpressionStatement)
	utils.Assert(t, ok)

	expr, ok := stmt.Expression.(T)
	utils.Assert(t, ok)
  return expr
}

func TestIdentifierExpression(t *testing.T) {
	input := "hello_World123"
  program := getProgram(t, input)

  ident := getExpr[*ast.Identifier](t, program)

	utils.AssertEq(t, ident.Name, input)
	utils.AssertEq(t, ident.Token,
		token.New(token.IDENTIFIER, input,
			token.NewSpan(0, 0, len(input)),
		),
	)
}

func TestIntegerExpression(t *testing.T) {
	input := "156_098"
	val := 156098

  program := getProgram(t, input)

  integer := getExpr[*ast.IntegerLiteral](t, program)

	utils.AssertEq(t, integer.Value, int64(val))
	utils.AssertEq(t, integer.Token,
		token.New(token.INTEGER, "156098",
			token.NewSpan(0, 0, len(input)),
		),
	)
}

func TestFloatExpression(t *testing.T) {
	input := "3.1415"
	val := 3.1415

  program := getProgram(t, input)

  float := getExpr[*ast.FloatLiteral](t, program)

	utils.AssertEq(t, float.Value, val)
	utils.AssertEq(t, float.Token,
		token.New(token.FLOAT, input,
			token.NewSpan(0, 0, len(input)),
		),
	)
}

func TestBooleanExpression(t *testing.T) {
	input := "true"

  program := getProgram(t, input)

  boolean := getExpr[*ast.BooleanLiteral](t, program)

	utils.AssertEq(t, boolean.Value, true)
	utils.AssertEq(t, boolean.Token,
		token.New(token.IDENTIFIER, input,
			token.NewSpan(0, 0, len(input)),
		),
	)
}

func TestErrorExpression(t *testing.T) {
  input := ")"

  l := lexer.New(input, "test.lb")
  tokens := l.Tokenise()
  p := parser.New(tokens, l.Diagnostics)
  program := p.Parse()

  utils.AssertEq(t, len(p.Diagnostics.Diagnostics), 1)
  diag := p.Diagnostics.Diagnostics[0]
  utils.AssertEq(t, diag.Message, "Expected expression, got `)`")
  utils.AssertEq(t, diag.Span, token.NewSpan(0, 0, 1))

  getExpr[*ast.ErrorExpression](t, program)
}

func TestMissingNewlineError(t *testing.T) {
  input := "1 2"

  l := lexer.New(input, "test.lb")
  tokens := l.Tokenise()
  p := parser.New(tokens, l.Diagnostics)
  p.Parse()

  utils.AssertEq(t, len(p.Diagnostics.Diagnostics), 1)
  diag := p.Diagnostics.Diagnostics[0]
  utils.AssertEq(t, diag.Message, "Expected newline after statement, got integer")
  utils.AssertEq(t, diag.Span, token.NewSpan(0, 2, 3))
}

