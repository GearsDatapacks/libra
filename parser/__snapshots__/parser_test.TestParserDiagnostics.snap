
[`)` - 1]
test.lb:1:1:
)
^ Expected expression, found `)`


---

[`let a = ;` - 1]
test.lb:1:9:
let a = ;
        ^ Expected expression, found `;`


---

[`1 2` - 1]
test.lb:1:3:
1 2
  ^ Expected newline after statement, found integer


---

[`(1 + 2` - 1]
test.lb:1:7:
(1 + 2
      ^ Expected `)`, found <Eof>


---

[`else; {}` - 1]
test.lb:1:1:
else
^ Else statement not allowed without preceding if


---

[`for i 42 {}` - 1]
test.lb:1:7:
for i 42 {}
      ^ Expected "in" keyword, found integer


---

[`let in = 1;for i in 20 {}` - 1]
test.lb:2:7:
for i in 20 {}
      ^ Expected "in" keyword, but it has been overwritten by a variable

test.lb:1:5:
let in = 1
    ^ Try removing or renaming this variable


---

[`fn add(a: i32, b, c): f32 {}` - 1]
test.lb:1:19:
fn add(a: i32, b, c): f32 {}
                  ^ The last parameter of a function must have a type annotation


---

[`fn foo(;bar;): baz {}` - 1]
test.lb:2:1:
bar
^ The last parameter of a function must have a type annotation

test.lb:1:4:
fn foo(
   ^ Parameter of this function


---

[`fn func_type(mut i32[]) {}` - 1]
test.lb:1:14:
fn func_type(mut i32[]) {}
             ^ "mut" must be followed by a parameter name


---

[`fn (string) bool.maybe() {}` - 1]
test.lb:1:13:
fn (string) bool.maybe() {}
            ^ Functions cannot be both methods and static members


---

[`import * from "io" as in_out` - 1]
test.lb:1:20:
import * from "io" as in_out
                   ^ Only one import modifier is allowed


---

[`import {read, write} from * from "io"` - 1]
test.lb:1:27:
import {read, write} from * from "io"
                          ^ Only one import modifier is allowed


---

[`if true { fn a() {} }` - 1]
test.lb:1:11:
if true { fn a() {} }
          ^ Function declaration not allowed here


---

[`type T = ;` - 1]
test.lb:1:10:
type T = ;
         ^ Expected type, found `;`


---

[`let value = .` - 1]
test.lb:1:14:
let value = .
             ^ Invalid right-hand side of expression. Expected identifier or struct body, found <Eof>


---

[`pub return 10` - 1]
test.lb:1:1:
pub return 10
^ Only top-level declarations can be exported


---

[`explicit fn func() {}` - 1]
test.lb:1:1:
explicit fn func() {}
^ Statement cannot be marked explicit


---

[`@nonexistent;fn attributed() {}` - 1]
test.lb:1:1:
@nonexistent
^ The attribute "nonexistent" does not exist


---

[`@tag FunctionTag;fn tagged() {}` - 1]
test.lb:2:1:
fn tagged() {}
^ Statement cannot be marked with attribute "tag"


---
